src/vulkan/engine_vulkan_debugger.cc

```diff
@@ -0,0 +1,976 @@
+// Copyright 2020 The Amber Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/vulkan/engine_vulkan.h"
+
+#if AMBER_ENABLE_VK_DEBUGGING
+
+#include <chrono>              // NOLINT(build/c++11)
+#include <condition_variable>  // NOLINT(build/c++11)
+#include <fstream>
+#include <mutex>  // NOLINT(build/c++11)
+#include <sstream>
+#include <thread>  // NOLINT(build/c++11)
+#include <unordered_map>
+
+#include "dap/network.h"
+#include "dap/protocol.h"
+#include "dap/session.h"
+
+// Set to 1 to enable verbose debugger logging
+#define ENABLE_DEBUGGER_LOG 0
+
+#if ENABLE_DEBUGGER_LOG
+#define DEBUGGER_LOG(...) \
+  do {                    \
+    printf(__VA_ARGS__);  \
+    printf("\n");         \
+  } while (false)
+#else
+#define DEBUGGER_LOG(...)
+#endif
+
+namespace {
+
+// TODO(bclayton): Make script configurable?
+static constexpr auto kThreadTimeout = std::chrono::minutes(1);
+
+// Event provides a basic wait-and-signal synchronization primitive.
+// TODO(bclayton): Move out to synchronization utility header?
+class Event {
+ public:
+  // Wait blocks until the event is fired.
+  void Wait() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&] { return signalled_; });
+  }
+
+  // Wait blocks until the event is fired, or the timeout is reached.
+  // If the Event was signalled, then Wait returns true, otherwise false.
+  template <typename Rep, typename Period>
+  bool Wait(const std::chrono::duration<Rep, Period>& duration) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return cv_.wait_for(lock, duration, [&] { return signalled_; });
+  }
+
+  // Signal signals the Event, unblocking any calls to Wait.
+  void Signal() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    signalled_ = true;
+    cv_.notify_all();
+  }
+
+ private:
+  std::condition_variable cv_;
+  std::mutex mutex_;
+  bool signalled_ = false;
+};
+
+// Split slices str into all substrings separated by sep and returns a vector of
+// the substrings between those separators.
+// TODO(bclayton): Move to string utility file?
```

Comment:
My usual thinking here is, if we have the same thing somewhere else, make ti common, otherwise leave it here for now and we can split it out later.

Reply:
Removed the TODO.